### PR TITLE
#461: Add Progressbar for Unpacking Archives

### PR DESCRIPTION
**Pull request by alfeilex**
Fixes: #461 

For archives that use the `extractArchive()` function a progressbar is shown during unpacking:

Affected Archives are:
- ZIP
- TAR (transitive PKG)
- JAR

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (alfeilex) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:617]: This workaround counting the size before init the progressbar works for ZIPs but JARs, as for example gcviewer.jar, have the size of -123

Probably that's still a bug in `extractJar()`

File: cli/src/test/java/com/devonfw/tools/ide/context/AbstractIdeTestContext.java
Comment (alfeilex) [inline cli/src/test/java/com/devonfw/tools/ide/context/AbstractIdeTestContext.java:91]: Not sure if that's wrong behavior, but if not put this condition then the tests for extracting tar files don't work. Probably because the tests run simultaniously? Could be `@BeforeEach` useful in this case?

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (alfeilex) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java]: progressbar for JAR extraction removed for now

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:670]: please remove `System.out`. If you want to keep it use logging on debug level.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:669]: did you test the overhead cause by this?
Does the read all data, unpack everything or will this just read some header/meta data and skip the rest of the payload?
As it seems the InputStream variants are slow in this regard.
We might need to implement this based on `ZipFile` instead:
https://stackoverflow.com/a/29331429/1506881

File: cli/src/test/java/com/devonfw/tools/ide/context/AbstractIdeTestContext.java
Comment (hohwille) [reply cli/src/test/java/com/devonfw/tools/ide/context/AbstractIdeTestContext.java:91]: Was added long time ago by @MustaphaOuchen 
I also would need to debug into this. Your change makes sense according to the existing workaround/hack.
However, IMHO the test progress bar should be improved but this can be done isolated from this PR so IMHO this change is fine and we can resolve this.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:703]: oh, nice one - kind of odd API. Thanks for finding out such tricky things.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:669]: Also if you do that workaround later for `TarArchiveEntry`, then IMHO the same has to be done here as well.

**Review by hohwille — COMMENTED** [review]
@alfeilex thanks for your PR. Great work. 👍 
On the long run, I would still suggest to create a generic abstraction of unpacking so we do not have to reimplement this redundantly for JAR, ZIP (when using JDK - see #552), etc.

My concern is still if the computation of the real size is causing a significant overhead and we should discuss this first. If that is clarified, we can approve but maybe the discussion will lead to some bigger changes/refactorings...

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (alfeilex) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java]: Sorry for that. I used that for testing and forgot to remove. 

done

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (alfeilex) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java]: Finally, there is not need for calculating the size of the archives. It's enough to take the file size. The issue before was probably that the ZIP extraction didn't work correctly and by increasing stepSize of the probgressbar the maximal size wasn't reached. Further, by using `ZipFile` we have the compressed size of entries and an overload iof the progressbar size doesn't occur anymore. 

Related issues: #575 and  #609

**Comment by alfeilex** [comment]
Blocked by #609

After the changes are merged, I will adapt the code and we can merge that PR after reviewing.

**Review by hohwille — APPROVED** [review]
@alfeilex thanks for your improvements. Cool idea to use the filesize and the compressed size of the entries what is a huge simplification with no real disadvantage. 👍

**Pull request merged**
